### PR TITLE
RM-45-records_mover-db-redshift-records_copy.py-13-1-C901-redshift_copy_options-is-too-complex-27

### PR DESCRIPTION
**Pull request by ryantimjohn**

### CODE REVIEW

File: records_mover/records/targets/fileobj.py [hunk 1]
```diff
@@ -74,8 +74,7 @@ def write_dfs(path_or_buf: Union[str, IO[str]]) -> int:
             move_count = write_dfs(text_fileobj)
             text_fileobj.detach()
         else:
-            with NamedTemporaryFile(prefix='mover_fileobj_target') as output_file:
-                move_count = write_dfs(output_file.name)
```
Comment (Brunope) [inline records_mover/records/targets/fileobj.py:78]: here's the reason your unit tests broken

Comment (ryantimjohn) [reply records_mover/records/targets/fileobj.py:78]: ty!

File: records_mover/db/redshift/records_copy.py [hunk 2]
```diff
@@ -3,92 +3,121 @@
...
-                          max_failure_rows: Optional[int]) -> RedshiftCopyOptions:
-    redshift_options: RedshiftCopyOptions = {}
+class RedshiftCopyParamaters:
+    def __init__(self,
+                 records_format: DelimitedRecordsFormat,
+                 fail_if_cant_handle_hint: bool,
+                 unhandled_hints: Set[str],
+                 fail_if_row_invalid: bool,
+                 max_failure_rows: Optional[int]):
+        self.redshift_options: RedshiftCopyOptions = dict()
```
Comment (Brunope) [inline records_mover/db/redshift/records_copy.py:21]: having both `RedshiftCopyParameters` and `RedshiftCopyOptions` is kinda confusing to me

File: records_mover/db/redshift/records_copy.py [hunk 3]
```diff
@@ -3,92 +3,121 @@
...
-    elif hints.compression == 'BZIP':
-        redshift_options['compression'] = Compression.bzip2
-    elif hints.compression is None:
-        redshift_options['compression'] = None
+def process_compression(redshift_copy_parameters: RedshiftCopyParamaters):
+    compression_map = {
+        'GZIP': Compression.gzip,
+        'LZO': Compression.lzop,
+        'BZIP': Compression.bzip2,
+        None: None
```
Comment (Brunope) [inline records_mover/db/redshift/records_copy.py:35]: Is this `None: None` necessary?

File: records_mover/db/redshift/records_copy.py [hunk 4]
```diff
@@ -3,92 +3,121 @@
...
 RedshiftCopyOptions = Dict[str, Optional[object]]
 
 
-def redshift_copy_options(unhandled_hints: Set[str],
-                          records_format: BaseRecordsFormat,
-                          fail_if_cant_handle_hint: bool,
-                          fail_if_row_invalid: bool,
-                          max_failure_rows: Optional[int]) -> RedshiftCopyOptions:
-    redshift_options: RedshiftCopyOptions = {}
+class RedshiftCopyParamaters:
```
Comment (Brunope) [inline records_mover/db/redshift/records_copy.py:14]: What do you think of using a factory abstraction instead? It seems like what we want to do is initialize _something_ with a few settings (fail_if_cant_handle_hint, fail_if_row_invalid, etc), then use that object to also store some state (unhandled_hints, copy_options) as we generate the desired output params.

Basically I think having a top-level object that stores these as `self` vars and has the `process_x` functions as methods could make a lot of sense. Thoughts?

**Comment by ryantimjohn** [comment]
converted back to job for the Factory pattern refactor

File: records_mover/db/redshift/records_copy.py (hunk 4 above)
Comment (ryantimjohn) [reply records_mover/db/redshift/records_copy.py:14]: I actually don't know that I'm going to write this as a factory pattern but we can look at it afterward and see if we need to refactor again.

File: records_mover/db/redshift/records_copy.py (hunk 2 above)
Comment (ryantimjohn) [reply records_mover/db/redshift/records_copy.py:21]: refactored

**Comment by ryantimjohn** [comment]
@Brunope just added docstrings after the approval so assuming it's okay to merge 😉

**Pull request merged**